src/lib/Microsoft.Health.Events/EventCheckpointing/StorageCheckpointClient.cs

```diff
@@ -92,49 +92,58 @@ public async Task UpdateCheckpointAsync(Checkpoint checkpoint)
             }
         }
 
-        public Task<Checkpoint> GetCheckpointForPartitionAsync(string partitionIdentifier, CancellationToken cancellationToken)
+        public async Task<Checkpoint> GetCheckpointForPartitionAsync(string partitionIdentifier, CancellationToken cancellationToken)
         {
             var prefix = $"{_blobPath}{partitionIdentifier}";
 
-            Task<Checkpoint> GetCheckpointAsync()
+            async Task<Checkpoint> GetCheckpointAsync()
             {
-                var checkpoint = new Checkpoint();
-
-                foreach (BlobItem blob in _storageClient.GetBlobs(traits: BlobTraits.Metadata, states: BlobStates.None, prefix: prefix, cancellationToken: cancellationToken))
+                try
                 {
-                    var partitionId = blob.Name.Split('/').Last();
+                    var requestConditions = new BlobRequestConditions();
+                    var propsResponse = await _storageClient.GetBlobClient(prefix).GetPropertiesAsync(requestConditions, cancellationToken);
```

Comment:
Also, do you know what states of blobs are returned by the blob client?  In the old code we kind of had a bug. The states variable was set to BlobStates.All which will also return soft deleted blobs.  This was causing issues when I was testing and wanted to reset my watermarks and start over from the beginning.  Deleting the blobs wasn't enough.  I just want to make sure we don't have a regression.

I checked the [BlobRequestConditions class](https://docs.microsoft.com/en-us/dotnet/api/azure.storage.blobs.models.blobrequestconditions?view=azure-dotnet) and it doesn't appear to have a similar filter there.

Reply:
> Just to confirm this is executed for a single blob? I guess the part that is confusing to me is variable name is prefix which would indicate this isn't the full blob path we are trying to match. Is that really the case?
> 
> Also, do we know if the blob client itself resolving the blob or is that deferred until you the GetPropertiesAsync method is called? Just wondering if there is a possible null reference error if the blob is resolved as part of the GetBlobClient call.

Yes, this will be executed for a single blob. The variable name was left over from the previous code. I'll rename it.

Looking at the source code for the client there is no interaction with the blob until a method is actually invoked on the client (`GetPropertiesAsync` in this case). It seems to be common pattern in the SDK to create a BlobClient and immediately invoke an operation. Here is an example where a [blob is deleted](https://github.com/Azure/azure-sdk-for-net/blob/3724ce24abd58311ec7846661870213b5f826911/sdk/storage/Azure.Storage.Blobs/src/BlobContainerClient.cs#L3130).

Reply:
> Also, do you know what states of blobs are returned by the blob client? In the old code we kind of had a bug. The states variable was set to BlobStates.All which will also return soft deleted blobs. This was causing issues when I was testing and wanted to reset my watermarks and start over from the beginning. Deleting the blobs wasn't enough. I just want to make sure we don't have a regression.
> 
> I checked the [BlobRequestConditions class](https://docs.microsoft.com/en-us/dotnet/api/azure.storage.blobs.models.blobrequestconditions?view=azure-dotnet) and it doesn't appear to have a similar filter there.

I had wondered the same. This method doesn't actually deal with the contents of the blob, but rather its properties. As such, I didn't see something like `BlobState` to actually pass in. Perhaps only the contents of a Blob are versioned, but there can only ever be one instance of its properties? Regardless, I'll check the source code to make sure that I am not missing something to avoid a regression.

Reply:
It looks like the Version of the returned blob needs to be specified as a part of the Url parameters that create the BlobClient. We could specify a new version using the `WithVersion` method on the returned `BlobClient`. But since we don't, we always get the latest version of the Blob.